File: packages/host-rest/src/keyed-mutex.ts
import { createKeyedMutex, type KeyedMutex } from "@kohaku-ui/host-core";
import type { KohakuHostDeps } from "./types.js";

// createKeyedMutex / KeyedMutex now live in host-core (the single implementation shared with the sample
// storage port and host-mcp-apps's fixation self-heal serialization; see host-core/src/keyed-mutex.ts for the
// doc comment). Re-exported here for backward compatibility with existing importers of this module.
export { createKeyedMutex, type KeyedMutex };

/**
 * A simple mutex that serializes the fixation's read-modify-write (self-healing's refreshFingerprint / invalidate
 * and the management plane's fixate / unfixate) per (tenant, intentHash) within the process (the same per-tenant
 * serialization strategy as the promotion lock).
 * Self-healing fires fire-and-forget from the compose path, so without serialization interleavings such as
 * "refreshFingerprint's get->put right after unfixate resurrects a deleted fixation" occur.
 * The lock is scoped per deps (= host instance) (in-process only; cross-process serialization is delegated to the
 * StoragePort's single-writer assumption).
 */
const fixationMutexByDeps = new WeakMap<KohakuHostDeps, KeyedMutex>();
export function withFixationLock<T>(
  deps: KohakuHostDeps,
  tenant: string | undefined,
  intentHash: string,
  fn: () => Promise<T>,
): Promise<T> {
  let mutex = fixationMutexByDeps.get(deps);
  if (mutex == null) {
    mutex = createKeyedMutex();
    fixationMutexByDeps.set(deps, mutex);
  }
  return mutex(`${tenant ?? ""} ${intentHash}`, fn);
}

File: python/kohaku/src/kohaku/host_core/keyed_mutex.py
"""Shared in-process keyed serialization lock (port of packages/host-core/src/keyed-mutex.ts).

Extracted from kohaku.host_rest's `_routes.shared` module (the original site of this implementation — see
that module's re-export of `_get_lock` / `_locks` for backward compatibility) so kohaku.host_mcp can consume
the identical mechanism for its fixation self-heal serialization, exactly as TS's `host-mcp-apps/src/server.ts`
consumes `@kohaku-ui/host-core`'s `createKeyedMutex` (the same relationship this module has with
`packages/host-rest/src/keyed-mutex.ts`'s re-export).

Unlike the TS port (a factory `createKeyedMutex()` returning a per-instance `KeyedMutex` closure over its own
`Map`), this is a single process-wide table keyed by `(event loop, owner, key)`: `owner` (typically a
`KohakuHostDeps` / `McpHostDeps` instance, compared via `id()`) plays the role TS's separate mutex-per-deps
`WeakMap` plays — scoping locks to one host instance so unrelated hosts sharing a process never contend on the
same lock — while the loop id makes this safe even for a test client that spans event loops (within the same
loop it is serialized).

Serializes the read-modify-write of promotion / fixation state across both host profiles: host_rest's
promotion lock and fixation lock, and host_mcp's fixation self-heal serialization (keyed by `intentHash`
alone, since the MCP profile never resolves a tenant). It only orders calls made *within this process*; see
`StoragePort`'s docstring (kohaku.spec) for the cross-process concurrency contract this mechanism sits on top
of.
"""

from __future__ import annotations

import asyncio
import contextlib
from collections.abc import AsyncIterator
from dataclasses import dataclass, field


@dataclass
class _LockEntry:
    """A serialization lock and its user count (held + waiting to acquire). Cleaned up from `_locks` when it
    reaches 0."""

    lock: asyncio.Lock = field(default_factory=asyncio.Lock)
    users: int = 0


_locks: dict[tuple[int, int, str], _LockEntry] = {}


@contextlib.asynccontextmanager
async def get_lock(owner: object, key: str) -> AsyncIterator[None]:
    """async context manager acquiring a per-(loop, owner, key) in-process serialization lock.

    Reference-counts the users and removes the entry from `_locks` when the last user leaves (without relying
    on asyncio.Lock private attributes). There is no await between get and increment (atomic under
    single-threaded asyncio), and waiters are also counted as users, so the entry is not cleaned up while any
    waiter remains.
    """
    loop_id = id(asyncio.get_running_loop())
    composite = (loop_id, id(owner), key)
    entry = _locks.get(composite)
    if entry is None:
        entry = _LockEntry()
        _locks[composite] = entry
    entry.users += 1
    try:
        async with entry.lock:
            yield
    finally:
        entry.users -= 1
        # If we are the last user (no waiters and no holder), clean up the entry.
        if entry.users == 0 and _locks.get(composite) is entry:
            del _locks[composite]
